src/i2s/std.rs

```diff
@@ -800,279 +743,85 @@ impl<'d, Dir: I2sRxSupported + I2sTxSupported> I2sStdDriver<'d, Dir> {
-        })
+        Self::internal_new_std(
+            i2s,
+            config,
+            true,
+            true,
+            bclk,
+            Some(din),
+            Some(dout),
+            mclk,
+            ws,
+        )
     }
+}
 
-    #[cfg(esp_idf_version_major = "4")]
+impl<'d> I2sDriver<'d, I2sRx> {
+    /// Create a new standard mode driver for the given I2S peripheral with only the receive channel open.
     #[allow(clippy::too_many_arguments)]
-    pub fn new_rx<I2S: I2s>(
-        _i2s: impl Peripheral<P = I2S> + 'd,
-        config: config::StdConfig,
+    pub fn new_std_rx<I2S: I2s>(
+        i2s: impl Peripheral<P = I2S> + 'd,
+        config: &config::StdConfig,
         bclk: impl Peripheral<P = impl InputPin + OutputPin> + 'd,
-        din: Option<impl Peripheral<P = impl InputPin> + 'd>,
+        din: impl Peripheral<P = impl InputPin> + 'd,
         mclk: Option<impl Peripheral<P = impl InputPin + OutputPin> + 'd>,
         ws: impl Peripheral<P = impl InputPin + OutputPin> + 'd,
     ) -> Result<Self, EspError> {
-        let port = I2S::port();
-        let mut driver_cfg = config.as_sdk();
-        driver_cfg.mode |= i2s_mode_t_I2S_MODE_RX;
-
-        // Safety: &driver_cfg is a valid pointer to an i2s_driver_config_t.
-        unsafe {
-            esp!(i2s_driver_install(port, &driver_cfg, 0, null_mut()))?;
-        }
-
-        // Set the pin configuration.
-        let pin_cfg = i2s_pin_config_t {
-            bck_io_num: bclk.into_ref().pin(),
-            data_in_num: din.map(|din| din.into_ref().pin()).unwrap_or(-1),
-            data_out_num: -1,
-            mck_io_num: mclk.map(|mclk| mclk.into_ref().pin()).unwrap_or(-1),
-            ws_io_num: ws.into_ref().pin(),
-        };
-
-        // Safety: &pin_cfg is a valid pointer to an i2s_pin_config_t.
-        unsafe {
-            esp!(i2s_set_pin(port, &pin_cfg))?;
-        }
-
-        Ok(Self {
-            i2s: port as u8,
-            _p: PhantomData,
-            _dir: PhantomData,
-        })
+        Self::internal_new_std(
+            i2s,
+            config,
+            true,
+            false,
+            bclk,
+            Some(din),
+            AnyIOPin::none(),
+            mclk,
+            ws,
+        )
     }
 }
 
-impl<'d, Dir: I2sTxSupported> I2sStdDriver<'d, Dir> {
+impl<'d> I2sDriver<'d, I2sTx> {
     /// Create a new standard mode driver for the given I2S peripheral with only the transmit channel open.
-    #[cfg(not(esp_idf_version_major = "4"))]
     #[allow(clippy::too_many_arguments)]
-    pub fn new_tx<I2S: I2s>(
-        _i2s: impl Peripheral<P = I2S> + 'd,
-        config: config::StdConfig,
+    pub fn new_std_tx<I2S: I2s>(
+        i2s: impl Peripheral<P = I2S> + 'd,
+        config: &config::StdConfig,
         bclk: impl Peripheral<P = impl InputPin + OutputPin> + 'd,
-        dout: Option<impl Peripheral<P = impl OutputPin> + 'd>,
+        dout: impl Peripheral<P = impl OutputPin> + 'd,
```

Comment:
Let's have this optional once DAC mode comes - even if that means a breaking change for the user. I think optionals are confusing, and should only be used if really necessary.

Reply:
Agreed. The more I look at it, I don't think this will ever happen, TBH. It's either buggy or very particular on the ESP32, and isn't supported on any other variant.